Mathlib/Tactic/Peel.lean

```diff
@@ -0,0 +1,215 @@
+      let (fvars, new_goal) ← and_imp.mvarId!.introN 2 [n.getD `_, n'.getD freshName]
+      return (fvars[1]!, [new_goal])
+    | (``Filter.Eventually, #[_, .lam _ _ b₁ _, f₁]),
+        (``Filter.Eventually, #[_, .lam n₂ t₂ b₂ c, f₂]) =>
+      unless ← isDefEq f₁ f₂ do
+        return (.none, [goal])
+      let all_imp ← mkFreshExprMVar <| ← withoutModifyingState <| withLocalDecl n₂ c t₂ fun x => do
+        mkForallFVars #[x] (← mkArrow (b₁.instantiate1 x) (b₂.instantiate1 x))
+      let event_forall ← mkAppOptM ``Filter.eventually_of_forall #[none, none, f₂, all_imp]
+      goal.assign (← mkAppM ``Filter.Eventually.mp #[e, event_forall])
+      let (fvars, new_goal) ← all_imp.mvarId!.introN 2 [n.getD n₂, n'.getD freshName]
+      return (fvars[1]!, [new_goal])
+    | (``Filter.Frequently, #[_, .lam _ _ b₁ _, f₁]),
+        (``Filter.Frequently, #[_, .lam n₂ t₂ b₂ c, f₂]) =>
+      unless ← isDefEq f₁ f₂ do
+        return (.none, [goal])
+      let all_imp ← mkFreshExprMVar <| ← withoutModifyingState <| withLocalDecl n₂ c t₂ fun x => do
+        mkForallFVars #[x] (← mkArrow (b₁.instantiate1 x) (b₂.instantiate1 x))
+      let event_forall ← mkAppOptM ``Filter.eventually_of_forall #[none, none, f₂, all_imp]
+      goal.assign (← mkAppM ``Filter.Frequently.mp #[e, event_forall])
+      let (fvars, new_goal) ← all_imp.mvarId!.introN 2 [n.getD n₂, n'.getD freshName]
+      return (fvars[1]!, [new_goal])
+    | _, _ =>
+      match ty, target with
+        | .forallE _ t₁ b₁ _, .forallE n₂ t₂ b₂ c => do
+          unless ← isDefEq (← whnfR t₁) (← whnfR t₂) do
+            return (.none, [goal])
+          let all_imp ← mkFreshExprMVar <| ← withoutModifyingState <|
+            withLocalDecl n₂ c t₂ fun x => do
+              mkForallFVars #[x] (← mkArrow (b₁.instantiate1 x) (b₂.instantiate1 x))
+          goal.assign (← mkAppM ``forall_imp #[all_imp, e])
+          let (fvars, new_goal) ← all_imp.mvarId!.introN 2 [n.getD n₂, n'.getD freshName]
+          return (fvars[1]!, [new_goal])
+        | _, _ => return (.none, [goal])
+
+/-- Peels `n` quantifiers off the expression `e` and the main goal without naming the introduced
+variables. The expression `e`, with quantifiers removed, is assigned the default name `this`. -/
+def peelNum (e : Expr) (n : Nat) : TacticM Unit := withMainContext do
+  match n with
+    | 0 => pure ()
+    | 1 => let _ ← liftMetaTacticAux (peelQuantifier · e (n' := `this))
+    | n + 2 =>
+      let fvar? ← liftMetaTacticAux (peelQuantifier · e (n' := `this))
+      if let some fvarId := fvar? then
+        peelNum (.fvar fvarId) (n + 1)
+        let mvarId ← (← getMainGoal).clear fvarId
+        replaceMainGoal [mvarId]
+
+/-- Given a list `l` of names, this continues to peel quantifiers off of the expression `e` and
+the main goal and introduces variables with the provided names until the list of names is exhausted.
+Note: the first name in the list is used for the name of the expression `e` with quantifiers
```

Comment:
I would suggest having `peelArgs` take another argument for this purpose, with type `Option Name`. This will simplify the `match` logic.

Reply:
I think it doesn't simplify the `match` logic very much, and it adds increased cost to the processing of the `withArgs`, so at best the complexity is only trading places. You're welcome to show me I'm wrong though! :smile: